src/NuGet.Core/NuGet.Protocol/Resources/PackageMetadataResourceV3.cs

```diff
@@ -38,37 +55,240 @@ public class PackageMetadataResourceV3 : PackageMetadataResource
-            parsed.PackageDetailsUrl = _packageDetailsUriResource?.GetUri(parsed.PackageId, parsed.Version);
-            return parsed;
+            var packageIdLowerCase = packageId.ToLowerInvariant();
+            var httpSourceCacheContext = HttpSourceCacheContext.Create(cacheContext, 0);
+
+            var index = await httpSource.GetAsync(
+                new HttpSourceCachedRequest(
+                    registrationUri.OriginalString,
+                    $"list_{packageIdLowerCase}_index",
+                    httpSourceCacheContext)
+                {
+                    IgnoreNotFounds = true,
+                },
+                async httpSourceResult => await processAsync(httpSourceResult),
+                log,
+                token);
+
+            return new RegistrationIndexResult
+            {
+                Index = index,
+                CacheContext = httpSourceCacheContext
+            };
+        }
+
+        /// <summary>
+        /// Process RegistrationIndex 
+        /// </summary>
+        /// <param name="httpSource">Httpsource instance.</param>
+        /// <param name="rangeUri">Paged registration index url address.</param>
+        /// <param name="packageId">PackageId for package we're checking.</param>
+        /// <param name="lower">Lower bound of nuget package.</param>
+        /// <param name="upper">Upper bound of nuget package.</param>
+        /// <param name="httpSourceCacheContext">SourceCacheContext for cache.</param>
+        /// <param name="log">Logger Instance.</param>
+        /// <param name="token">Cancellation token.</param>
+        /// <returns></returns>
+        private async Task<RegistrationPage> GetRegistratioIndexPageAsync(
+            HttpSource httpSource,
+            string rangeUri,
+            string packageId,
+            NuGetVersion lower,
+            NuGetVersion upper,
+            HttpSourceCacheContext httpSourceCacheContext,
+            ILogger log,
+            CancellationToken token)
+        {
+            var packageIdLowerCase = packageId.ToLowerInvariant();
+            var registrationPage = await httpSource.GetAsync(
+                            new HttpSourceCachedRequest(
+                                rangeUri,
+                                $"list_{packageIdLowerCase}_range_{lower.ToNormalizedString()}-{upper.ToNormalizedString()}",
+                                httpSourceCacheContext)
+                            {
+                                IgnoreNotFounds = true,
+                            },
+                            async httpSourceResult =>
+                            {
+                                return await ProcessRegistrationIndexAsync<RegistrationPage>(httpSourceResult, token);
+                            },
+                            log,
+                            token);
+
+            return registrationPage;
+        }
```

Comment:
The method only has a single `await`, and it's the last thing in the method, therefore we can avoid memory allocations by not needing an async state machine.

```suggestion
        private Task<RegistrationPage> GetRegistratioIndexPageAsync(
            HttpSource httpSource,
            string rangeUri,
            string packageId,
            NuGetVersion lower,
            NuGetVersion upper,
            HttpSourceCacheContext httpSourceCacheContext,
            ILogger log,
            CancellationToken token)
        {
            var packageIdLowerCase = packageId.ToLowerInvariant();
            var registrationPage = httpSource.GetAsync(
                            new HttpSourceCachedRequest(
                                rangeUri,
                                $"list_{packageIdLowerCase}_range_{lower.ToNormalizedString()}-{upper.ToNormalizedString()}",
                                httpSourceCacheContext)
                            {
                                IgnoreNotFounds = true,
                            },
                            async httpSourceResult =>
                            {
                                return await ProcessRegistrationIndexAsync<RegistrationPage>(httpSourceResult, token);
                            },
                            log,
                            token);

            return registrationPage;
        }
```

Reply:
This was good suggestion, I didn't realize it.